### ISSUE TITLE
Define a transform on API JSON data which produces unix-friendly, greppable output

### ISSUE THREAD

**Issue by sirosen** [globus/globus-cli#403]
Example of a task we had in a support ticket:
- Safely extract the endpoint ID of `foo#bar` (canonical name) from `globus endpoint search` results

Naive approach:
```
globus endpoint search --filter-owner-id foo@globusid.org 'foo#bar' | \
    tail -n1 | cut -f1
```

But this doesn't work for a few reasons. It's not guaranteed 1 result, for starters, and it doesn't guarantee ordering of results.

Current (real) solution:
```
globus endpoint search --filter-owner-id foo@globusid.org 'foo#bar' \
    --jmespath 'DATA[?canonical_name==`foo#bar`]|[0].id' | tr -d '"'
```

But what the user wants:
```
globus endpoint search --filter-owner-id foo@globusid.org 'foo#bar' --output unix | \
    awk '$2=="foo#bar"{print $1}'
```
(or similar. The point is that awk, grep, and other tools can be used to do the processing safely.)

The trick is to define something which
- Works all the time, even when given data which is not perfectly symmetric
- Provides unix-friendly output
- Correctly handles encoding of quotes and whitespace characters, with robustness to confidently handle (perhaps in a degraded way) new API responses which may contain "bad" data
- Retains identifiers for datadicts when necessary, but drops them when columnar formatting is possible?
  e.g. `[{"foo": "bar"}, {"foo": "baz"}]` can become
  ```
  foo
  bar
  baz
  ```
  or even
  ```
  bar
  baz
  ```
  but `[{"foo": "bar"}, {"foo": "baz"}, {"fizz": "buzz"}]` either needs to render as
  ```
  foo fizz
  bar none
  baz none
  none buzz
  ```
  or otherwise as
  ```
  foo bar
  foo baz
  fizz buzz
  ```

- Isn't perilous for us to maintain long term
- Can support composition with `--jmespath` to allow jmespath query capabilities pipelined with traditional unix commands
- Works in a predictable, though potentially non-obvious, way on user-defined JSON data
- Defines explicit exceptions to safe formatting rules in the interests of expediency and usability
  e.g. Rendering `"true"` and `true` identically, or `null` and `"null"` identically may be worth it for a more usable tool. If we produce TSV data, rendering `\t` in a data field is not okay.

All of this points in the direction of being truly general with respect to the input document, but not necessarily enforcing a really pedantic notion of "safety" on users at the cost of usability. You can always increase the safety of your results by explicitly selecting keys with `--jmespath`, and this is in fact the approach advocated by the awscli for its `--query '...' --output text` mode.

**Comment by sirosen** [comment]
#404 is an approach which handles all of these constraints except for

> Works all the time, even when given data which is not perfectly symmetric

As noted there, it can fail on certain data structures, but more explicit querying can remedy the situation.
I think this is an acceptable fault if it meets all of these other criteria.

**Comment by sirosen** [comment]
Closed by #404
